### PR TITLE
Conditional imputer

### PR DESCRIPTION
**Pull request by satya-pattnaik**
#### Reference Issues/PRs

Can be used with detectors from #798 

#### What does this implement/fix? Explain your changes.

A Conditional Imputer which takes a annotator/detector, imputer and conditionally imputes the annotated data points using Sktime Imputer. 
#### What should a reviewer concentrate their feedback on?

Codea nd directory structure of the ConditionalImputer.
##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Comment by aiwalter** [comment]
@satya-pattnaik could we make this also working for having `Z` as a `pd.DataFrame`? I recently enabled this for `Imputer` already..

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+       impute values set to nan by the detector.
+    Parameters
+    ----------
+    imputer : Sktime Imputer
+    detector : Sktime Annotator/Detector
+    """
+
+    _tags = {"skip-inverse-transform": True}
+
+    def __init__(self, imputer, detector):
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:23]:
```suggestion
    def __init__(self, imputer, annotator):
```

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+class ConditionalImputer(_SeriesToSeriesTransformer):
+
+    """Conditonal Imputer that uses a Sktime Imputer to
+       impute values set to nan by the detector.
+    Parameters
+    ----------
+    imputer : Sktime Imputer
+    detector : Sktime Annotator/Detector
+    """
+
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:20]:
```suggestion
_required_parameters = ["imputer", "annotator"]
```

Comment (mloning) [reply sktime/transformations/series/compose/_conditional_impute.py:20]: add the `ConditionalImputer` to the config in `sktime/tests/_config.py`

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        X : pd.DataFrame, optional (default=None)
+            Exogenous variables are ignored
+        Returns
+        -------
+        self : returns an instance of self.
+        """
+
+        z = check_series(Z, enforce_univariate=True)
+
+        self.detector_ = self.detector.fit(z)
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:43]:
```suggestion
        self.annotator_ = clone(self.annotator)
        self.annotator_.fit(z)
```

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+        z = z.to_numpy().reshape(-1, 1)
+        self.imputer_ = self.imputer.fit(z)
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:46]: use `clone` (see above)

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        Returns
+        -------
+        self : returns an instance of self.
+        """
+
+        z = check_series(Z, enforce_univariate=True)
+
+        self.detector_ = self.detector.fit(z)
+
+        z = z.to_numpy().reshape(-1, 1)
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:45]:
```suggestion
```

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            Exogenous variables are ignored
+        Returns
+        -------
+        Zt : pd.Series
+            Transformed(Conditionally Imputed) time series.
+        """
+
+        z = check_series(Z, enforce_univariate=True)
+
+        outliers = self.detector_.transform(z)
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:66]:
```suggestion
        outliers = self.annotator_.predict(z)
```

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+        Zt = Z.copy()
+        Zt.iloc[outliers[outliers].index] = np.nan
+
+        z = Zt
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:71]:
```suggestion  
```

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        -------
+        Zt : pd.Series
+            Transformed(Conditionally Imputed) time series.
+        """
+
+        z = check_series(Z, enforce_univariate=True)
+
+        outliers = self.detector_.transform(z)
+
+        Zt = Z.copy()
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:68]:
```suggestion
```

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        Zt.iloc[outliers[outliers].index] = np.nan
+
+        z = Zt
+        zt = self.imputer_.transform(z)
+        return zt
+
+    def fit_transform(self, Z, X=None):
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:75]: in the base class we have: `fit(...).transform(...)`

**Review by mloning — COMMENTED** [review]
As discussed, here are our notes.

In compositions, the changing output type of the predict method from annotators leads to tricky case distinctions (similar to what we have seen with prediction intervals in forecasting). Possible solutions
* allow us to query `fmt`/`labels` settings
* always return some standard `fmt` and `label` setting when calling predict
* quick fix: enforce certain settings

```python
def check_annotator(annotator):
    assert annotator.get_params("fmt") == "dense"
    assert annotator.get_params("labels") == "bool"
```

**Comment by fkiraly** [comment]
>In compositions, the changing output type of the predict method from annotators leads to tricky case distinctions (similar to what we have seen with prediction intervals in forecasting). Possible solutions

Why that?

The fourth option, not listed, is in my opinion the obvious one.
Annotators implement both interfaces, so when initializing the estimator as a composite, just initialize the annotator with the one you want. That is,

```python

self.annotator_ = clone(annotator)
self.annotator_.set_params({"fmt" : "dense"})
```

Etc. Alternatively, we could move the settings to the `predict` method, but I'm not sure whether it's sensible to have it passed that late.

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 10]
```diff
@@ -40,10 +43,12 @@ def fit(self, Z, X=None):
...
 
-        self.detector_ = self.detector.fit(z)
+        self.annotator_ = clone(self.annotator)
+        self.annotator_.set_params({"fmt": "dense"})
+        self.annotator_.fit(z)
 
-        z = z.to_numpy().reshape(-1, 1)
-        self.imputer_ = self.imputer.fit(z)
+        self.imputer_ = clone(self.imputer)
+        self.imputer_.fit(z)
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:51]:
```suggestion
        self.imputer_.fit(z, X)
```

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 11]
```diff
@@ -63,40 +68,8 @@ def transform(self, Z, X=None):
...
-
-        z = check_series(Z, enforce_univariate=True)
-
-        self.detector_ = self.detector.fit(z)
-
-        outliers = self.detector_.transform(z)
-
-        Zt = Z.copy()
-        Zt[outliers[outliers].index] = np.nan
+        outliers = self.annotator_.transform(z)
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:71]:
```suggestion
        outliers = self.annotator_.transform(z, X)
```

File: sktime/transformations/series/compose/_conditional_impute.py [hunk 12]
```diff
@@ -63,40 +68,8 @@ def transform(self, Z, X=None):
...
 
-        z = Zt
-        self.imputer_ = self.imputer.fit(z)
-        zt = self.imputer_.fit_transform(z)
-        return zt
+        z.iloc[outliers[outliers].index] = np.nan
+        z = self.imputer_.transform(z)
```
Comment (mloning) [inline sktime/transformations/series/compose/_conditional_impute.py:74]:
```suggestion
        z = self.imputer_.transform(z, X)
```
We probably also have to subset the X as we do with z above.

**Review by mloning — COMMENTED** [review]
Hi @satya-pattnaik - just a few quick comments!

**Review by mloning — COMMENTED** [review]
Hi @satya-pattnaik, I think this is ready to go. I just have to minor comments which we could address before merging.

**Comment by mloning** [comment]
We also need to add it to the online docs.

**Comment by RNKuhns** [comment]
@satya-pattnaik and @mloning where does this stand? Is there anything I can to help get this merged?

**Comment by satya-pattnaik** [comment]
> @satya-pattnaik and @mloning where does this stand? Is there anything I can to help get this merged?

Apologies for late reply, I was OOO. I am working on this , will give you the update by Monday(8 Nov).
